File: 코드/ch8/clip3/boj24268/Main.java
import java.util.Scanner;

public class Main {
    public static int n;
    public static int d;
    public static boolean[] usedNumber;
    public static int[] number;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        d = sc.nextInt();

        usedNumber = new boolean[d];
        number = new int[d];
        int result = findPermutation(0);
        System.out.println(result);
    }

    public static int findPermutation(int depth) {
        if (depth == d) {
            // 3. 진법변환
            int result = 0;
            for (int i = 0; i < depth; i++) {
                result *= d;
                result += number[i];
            }
            return result;
        }

        // 2. 앞자리가 0이면 무시하도록
        int start = 0;
        if (depth == 0) start = 1;
        for (int i = start; i < d; i++) {
            // 1. 중복되는 수가 생성되지 않도록
            if (!usedNumber[i]) {
                usedNumber[i] = true;
                number[depth] = i;
                int num = findPermutation(depth + 1);
                if (num > n) return num;

                usedNumber[i] = false;
            }
        }
        return -1;
    }
}

File: 코드/ch8/clip3/boj24268/main.py
import sys

def find_permutation(depth):
    if depth == d:
        result = 0
        for i in range(depth):
            result *= d
            result += number[i]
        return result

    start = 0
    if depth == 0:
        start = 1
    for i in range(start, d):
        if not used_number[i]:
            used_number[i] = True
            number[depth] = i
            num = find_permutation(depth + 1)
            if num > n:
                return num
            used_number[i] = False
    return -1

n, d = map(int, sys.stdin.readline().split())
used_number = [False] * d
number = [0] * d

result = find_permutation(0)
sys.stdout.write(str(result) + '\n')
